packages/core/causalorder/causalorder.go

```diff
@@ -192,67 +161,82 @@ func (c *CausalOrder[ID, Entity]) popUnorderedChildren(entityID ID) (pendingChil
 	return pendingChildren
 }
 
-func (c *CausalOrder[ID, Entity]) propagateCausalOrder(metadata Entity) {
-	for childWalker := walker.New[ID](true).Push(metadata.ID()); childWalker.HasNext(); {
-		for _, child := range c.popUnorderedChildren(childWalker.Next()) {
-			if c.triggerChildIfReady(child) {
-				childWalker.Push(child.ID())
-			}
-		}
+func (c *CausalOrder[ID, Entity]) triggerChildIfReady(child Entity) {
+	c.dagMutex.Lock(child.ID())
+	defer c.dagMutex.Unlock(child.ID())
+
+	if c.decreaseUnorderedParentsCounter(child) == 0 {
+		c.triggerOrderedCallback(child)
 	}
 }
 
-func (c *CausalOrder[ID, Entity]) triggerChildIfReady(child Entity) (eventTriggered bool) {
-	c.dagMutex.Lock(child.ID())
-	defer c.dagMutex.Unlock(child.ID())
+func (c *CausalOrder[ID, Entity]) triggerOrderedCallback(entity Entity) (wasTriggered bool) {
+	if err := c.orderedCallback(entity); err != nil {
+		c.droppedCallback(entity, err)
 
-	if c.decreaseUnorderedParentsCounter(child) != 0 || !c.setOrdered(child) {
-		return false
+		return
 	}
 
-	c.Events.Emit.Trigger(child)
+	c.propagateOrderToChildren(entity.ID())
 
 	return true
 }
 
-func (c *CausalOrder[ID, Entity]) decreaseUnorderedParentsCounter(metadata Entity) (unorderedParentsCounter uint8) {
-	c.unorderedParentsCounterMutex.Lock()
-	defer c.unorderedParentsCounterMutex.Unlock()
+func (c *CausalOrder[ID, Entity]) propagateOrderToChildren(id ID) {
```

Comment:
Does it need to be a separate method? It's only used once in a method above and it's not too complex.

Reply:
+1